module/validation/receipt_validator.go

```diff
@@ -172,16 +175,29 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
 // 	* execution result has a valid parent
 // Returns nil if all checks passed successfully
 func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
+	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
+	// local storage. This is needed to handle a case when same block payload contains receipts that
+	// reference each other.
+	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
+		for _, receipt := range receipts {
+			if executionResult.PreviousResultID == receipt.ExecutionResult.ID() {
+				return &receipt.ExecutionResult, nil
+			}
+		}
+
+		return v.previousResult(executionResult)
+	}
+
 	for i, r := range receipts {
-		err := v.validate(r)
+		err := v.validate(r, previousResult)
 		if err != nil {
 			return fmt.Errorf("could not validate receipt %v at index %v: %w", r.ID(), i, err)
 		}
 	}
 	return nil
 }
 
-func (v *receiptValidator) validate(receipt *flow.ExecutionReceipt) error {
+func (v *receiptValidator) validate(receipt *flow.ExecutionReceipt, previousResult GetPreviousResult) error {
```

Comment:
I don't think this works. 

For instance, 
```
A <- B(A) <- C <- D
```

If I'm building such block E: `D<-E(B, C, D)`, then this logic would consider `D` as invalid, because the previous result `C` doesn't exist in storage yet.